Establish updated heatmap styling (#531)
AlexsLemonade/scpca-nf: templates/qc_report/celltypes_supplemental_report.rmd

celltype_df <- create_celltype_df(processed_sce)
```

> ## Comparison to clusters

Reviewer: Here is my stab at how to organize the text with subsections. 

```
## Cell label comparison plots

This section displays heatmaps comparing cell type labels from various methods. 

We use the [Jaccard similarity index](https://en.wikipedia.org/wiki/Jaccard_index) to display the agreement between between pairs of labels from different methods.

...

### Comparison with unsupervised clustering

Here we show the results from unsupervised clustering compared against cell type annotations. 
Cluster assignment was performed using the `r metadata(processed_sce)$cluster_algorithm` algorithm.
```
(I left off weighting here: I think that is more detail than we need in this report, and it does not capture the full array of parameters which would really be needed to compare to other clustering results)

Reviewer 2: +1 for this! This is similar to what I had envisioned when I wrote that comment.

Reviewer 2: Another thought, we could tackle the heatmap aesthetics in this PR, turn it into a function in the next PR, and then add in the text as a separate PR. That might help focus the remaining reviews.

Author: > Another thought, we could tackle the heatmap aesthetics in this PR, turn it into a function in the next PR, and then add in the text as a separate PR. That might help focus the remaining reviews.

yeah, i'm good with this. I want to get this text in now, but we can polish it up later!

## Heatmaps

> Below, we show heat maps comparing each set of cell type annotations to clustering results.

Reviewer: I added "unsupervised" here, but I actually think part of this sentence needs to be moved down, as we want to start general and then have this sentence for the first set of heatmaps, then a different one for the ones below (comparing to submitter, etc.)

```suggestion
Below, we show heat maps comparing each set of cell type annotations to unsupervised clustering results.
```

> Heatmap colors represent the Jaccard similarity index between each grouping.
> 
> This metric is bounded by 0 and 1:
> 
> * A value of 1 indicates complete overlap between categories.
> The categories being compared contain the same set of cells.
> * A value of 0 indicates no overlap between categories. 
> No cells that appear in one category appear in the other.

Reviewer: I'm trying here to clarify that the comparisons are always between labels in different methods, but since they are not all cell typing methods, it is a bit hard to word. Here is an attempt.

```suggestion
Heatmap colors represent the Jaccard similarity index between pairs of labels from different methods.
The Jaccard index reflects the degree of overlap between the two labels, and ranges from 0 to 1. 

* If the labels are assigned to identical sets of cells, the Jaccard index will be 1
* If the labels are assigned to completely non-overlapping sets of cells, the Jaccard index will be 0
* 
```

> No cells that appear in one category appear in the other.

Reviewer 2: I think I might reorganize this a bit. We are using the Jaccard index for both this heatmap and all the other heatmaps right? So I think I would start with a general statement about the section as a whole - that we are interested in comparing cell type annotations to clusters and to each other. Then I would say that we are using the Jaccard Index to measure the similarity between these different groups. Then you have an intro sentence before each heatmap saying what exactly is being compared and used to calculate the Jaccard index for that heatmap.

keep_legend_name <- names(jaccard_cluster_matrices)[1]

# Set heatmap padding option
ComplexHeatmap::ht_opt(TITLE_PADDING = grid::unit(0.2, "in"))

# TODO: make this into a function so can re-use for 2nd group of heatmaps (still forthcoming)
heatmap_list <- jaccard_cluster_matrices |> 
  purrr::imap(\(mat, name){
    ComplexHeatmap::Heatmap(
      t(mat), # transpose because matrix rows are in common & we want a vertical arrangement
      col = col_fun,
      border = TRUE, # each heatmap gets its own outline
      ## Row parameters
      cluster_rows = FALSE,
      row_title = name, # each heatmap gets its own title
      row_title_gp = grid::gpar(fontsize = 10), 
      row_title_side = "right",
      row_names_side = "left",
      row_names_gp = grid::gpar(fontsize = 8),
      ## Column parameters
      cluster_columns = FALSE,
      column_title = glue::glue("Cluster"),
      column_names_side = "top",
      column_names_rot = 0,
      column_names_gp = grid::gpar(fontsize = 8), # match row name size
      ## Legend parameters
      heatmap_legend_param = list(
        title = "Jaccard similarity index",
        direction = "horizontal",

>         legend_width = unit(4, "cm")

Reviewer: For consistent units
```suggestion
        legend_width = unit(1.5, "in")
```

      ),
      # Retain only 1 legend in the final plot      
      show_heatmap_legend = name == keep_legend_name,
    )
  }) |>
  # concatenate vertically into HeatmapList object
  purrr::reduce(ComplexHeatmap::`%v%`) 

# Render the heatmap list
ComplexHeatmap::draw(
  heatmap_list, 
  heatmap_legend_side = "bottom" 
)
```

```{r, eval = has_submitter & (has_cellassign | has_singler)}
methods_string <- dplyr::case_when(

>   has_cellassign & has_singler ~ glue::glue("`SingleR` and `CellAssign`"),
>   has_cellassign  ~ glue::glue("`CellAssign`"),
>   has_singler  ~ glue::glue("`SingleR`"),

Reviewer: I don't think `glue` is doing anything here?

(Also made order consistent)
```suggestion
  has_singler & has_cellassign ~ "`SingleR` and `CellAssign`",
  has_singler  ~ "`SingleR`",
  has_cellassign  ~ "`CellAssign`",
```

Author: > I don't think `glue` is doing anything here?

I thought it might be good for printing these as code-styled when used elsewhere, in particular since one has an "and" which we don't want to be code-style, but maybe if it's inside a `glue::glue()` later it will work without that here! Will test and Do the Right Thing™️.

    "cellassign_celltype_annotation"
  )
}

heatmap_height <- length(jaccard_cluster_matrices) * 2.5
```

```{r, fig.height = heatmap_height, fig.width = 7, warning = FALSE}
# Define color ramp for shared use in the heatmap - TO BE DISCUSSED!

> col_fun <- circlize::colorRamp2(c(0, 1), colors = c("yellow", "red"))

Reviewer: Since we don't have NA values (I don't think we can?) I would probably favor starting with white and going to a nice dark blue or purple here. Something like what is shown in the bottom plot here: https://towardsdatascience.com/consensus-clustering-f5d25c98eaf2

Reviewer 2: I like this idea of using white for close to 0 and a dark color for 1.

Author: > Since we don't have NA values (I don't think we can?) 

Confirming we can't, 🐛's aside! All the `NA`s will have been converted at this point to strings.

Reviewer: I meant for the Jaccard calculation, which better not be strings!

Author: they better not! 😱

# We only want one shared legend in the end, arbitrarily grab the first one
keep_legend_name <- names(jaccard_cluster_matrices)[1]

###############################################################################
# BELOW FAILS WITH ERROR:  - Error: object 'ComplexHeatmap' not found...????

> #ComplexHeatmap::ht_opt[["TITLE_PADDING"]]<- unit(0.5, "cm")

Reviewer: Did you try this form?
```suggestion
ComplexHeatmap::ht_opt(TITLE_PADDING = grid::unit(0.5, "cm"))
```

Reviewer: A secondary comment here: `fig.height` is in inch units so maybe we should stick with that throughout?

# BELOW WORKS!
suppressPackageStartupMessages(library(ComplexHeatmap))
ht_opt[["TITLE_PADDING"]] <- unit(0.5, "cm")
###############################################################################

# TODO: make this into a function so can re-use for 2nd group of heatmaps (still forthcoming)
heatmap_list <- jaccard_cluster_matrices |> 
  purrr::imap(\(mat, name){
    ComplexHeatmap::Heatmap(
      t(mat), # transpose because matrix rows are in common & we want a vertical arrangement
      col = col_fun,
      ## Row parameters
      cluster_rows = FALSE,
      row_title = name, # each heatmap gets its own title
      row_title_gp = grid::gpar(fontsize = 10), 
      row_title_side = "right",
      row_names_side = "left",
      row_names_gp = grid::gpar(fontsize = 8),
      ## Column parameters
      cluster_columns = FALSE,
      column_title = glue::glue("Cluster"),
      column_names_side = "top",
      column_names_rot = 0,

>       column_names_gp = grid::gpar(fontsize = 12),

Reviewer: Even though we have more space here, I think probably we want to keep just one font size for rows and columns?

      ## Legend parameters
    heatmap_legend_param = list(
      title = "Jaccard similarity index",
      direction = "horizontal",
      legend_width = unit(4, "cm")
    ),
    # Retain only 1 legend in the final plot      

>     show_heatmap_legend = ifelse(name == keep_legend_name, TRUE, FALSE),

Reviewer: No need for `ifelse` if just T/F!
```suggestion
    show_heatmap_legend = name == keep_legend_name,
```

Author: The sheer number of times I have corrected something like this on homeworks....for me to go do it myself is...impressive!
